### ISSUE TITLE
Cannot call functions from inside anonymous function in EM_ASM

### ISSUE THREAD

**Issue by Volkhyn** [kripken/emscripten#4925]
Hello,
I'm not sure that this is an issue or if I missed something in the documentation but If I call functions (even basic ones as alert) from an anonymous function inside EM_ASM_ARGS I got this error : 

```
C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:272
        throw new JS_Parse_Error(message, line, col, pos);
        ^
TypeError: (intermediate value) is not a function
    at new JS_Parse_Error (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:261:21)
    at js_error (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:272:15)
    at croak (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:742:17)
    at token_error (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:749:17)
    at unexpected (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:755:17)
    at Object.semicolon (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:775:51)
    at prog1 (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:1314:29)
    at simple_statement (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:906:35)
    at C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:816:35
    at block_ (C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:999:32)
    at C:\Program Files (x86)\ShiVa_Technologies\ShiVa_Editor\Tools\Windows\emsdk\emscripten\1.35.0\tools\eliminator\node_modules\uglify-js\lib\parse-js.js:821:52`
```

Here is my code : 
```
`extern "C" {
		void EMSCRIPTEN_KEEPALIVE fillRequestResult(int id,const char * result, const char * requestType, const char * URL, const char * dataType, const char * dataToPost, const char * AIName, const char * EventName, int statusCode) {
			RequestData request;
			request.id = id;
			request.result = result;
			request.statusCode = statusCode;
			request.URL = URL;
			request.requestType = requestType;
			request.dataToPost = dataToPost;
			request.dataType = dataType;
			request.AIName = AIName;
			request.EventName = EventName;

			::httpResults.push_back(request);
		}
	}

	/*
	* Name : sendRequest
	* Description : Send the HTTP request
	* Parameters
	* iRequest : request container
	*/
	void sendRequest(RequestData& iRequest)
	{
		EM_ASM_ARGS({

			var xhr = new XMLHttpRequest();
			xhr.open(Pointer_stringify($0), Pointer_stringify($1), true);
			if (Pointer_stringify($0).localeCompare('PUT')) {
				xhr.send();
			}
			else {
				xhr.send();
			}

			xhr.onreadystatechange = function() {
 //Cannot call anything from here without errors
				if (xhr.readyState == 4 && xhr.status == 200) {
					if (Pointer_stringify($2).search('XML') != -1) {
						Module.Runtime.dynCall('v', $7, [$4, xhr.responseText, Pointer_stringify($0), Pointer_stringify($1), Pointer_stringify($2), Pointer_stringify($3), Pointer_stringify($5), Pointer_stringify($6), xhr.status]);
					}
					else if (Pointer_stringify($2).search('JSON') != -1) {
						else if (Pointer_stringify($2).search('QA') != -1) {

						}
					}
				}
			}

		}, iRequest.requestType.c_str(), iRequest.URL.c_str(), iRequest.dataType.c_str(), iRequest.dataToPost.c_str(), iRequest.id, iRequest.AIName.c_str(), iRequest.EventName.c_str(), &fillRequestResult);

	}`
```

**Comment by kripken** [comment]
I'm not sure, but I suspect this is hitting a limitation of the C preprocessor. C does process the EM_ASM contents, and some patterns confuse it (like array literals, I think I remember). I would try to remove code until you find what is the critical line.

**Comment by Volkhyn** [comment]
Evrything put in the anonymous function will throw this TypeError: (intermediate value) is not a function.

So this compiles : 
```
xhr.onreadystatechange = function() {

}
```

This won't : 
```
xhr.onreadystatechange = function() {
	if (xhr.readyState == 4 && xhr.status == 200) {
				
	}
}
```

This won't compile too : 
```
xhr.onreadystatechange = function() {
	alert('Hey!');
}
```

**Comment by kripken** [comment]
I see. Yeah, probably a C preprocessor limitation, then, with nested `{`.

To avoid this, you can write your JS in a `--js-library` or `--pre-js`, and just call it from the EM_ASM.

**Comment by Volkhyn** [comment]
I wanted to try what you suggested but I ran into an issue with the library linking.
I created a js file library_httpRequest.js where I put this test case taken from the documentation : 
```
mergeInto(LibraryManager.library, {
    test: function() {
        alert('hi');
    },
});
```
and changed the C++ file for : 
```
extern "C" {
		extern void EMSCRIPTEN_KEEPALIVE test(void);
	}

	/*
	* Name : sendRequest
	* Description : Send the HTTP request
	* Parameters
	* iRequest : request container
	*/
	void sendRequest(RequestData& iRequest)
	{
		test();
        }
```

I also added the option 
`-s EXPORTED_FUNCTIONS="['_test']" --js-library "C:\Users\cpignol\Documents\Incollables\Charlotte\Plugins\com.gayatech.networkHandler\Sources\HTML\library_httpRequest.js"`

to my makefile but I got an error "unresolved symbol: test".

I also tried EM_ASM(test();); and to put _ before the function name but still, same error. What could I have done wrong ?

**Comment by kripken** [comment]
A JS library defines a C API. So you would need to call `test` from the C code.

If you want to call it from JS, then a pre-js is better. It just prepends the pre-js, and then optimizes it together so everything should work.

**Comment by Volkhyn** [comment]
I think I tried all I can, I wrapped in an extern C the function that call the js : 
```
extern "C" {
    extern void EMSCRIPTEN_KEEPALIVE test();
    void sendRequest(){
	test();
     }
}
```
and used 
`-s EXPORTED_FUNCTIONS="['_test','_sendRequest']" --pre-js "C:\Users\cpignol\Documents\Incollables\Charlotte\Plugins\com.gayatech.networkHandler\Sources\HTML\library_httpRequest.js"`
Again, tried with and without _ and inside EM_ASM without success.
I also tried to add a postset line : 
```
mergeInto(LibraryManager.library, {
    test__postset: '_test();',
    test: function() {
        alert('hi');
    },
});
```
but still this missing "test" function error.
I don't even have any log related to a linking process with the js library file (I assume I should not see it appear only in the log printing the command line)

**Comment by kripken** [comment]
When you call it from C (using extern C, which is correct), then it must be a js-library. JS libraries are just like libc or other system libs, they are C APIs that are linked in on demand. So we must see a C call to them.

When you call it from JS (which is what a call from an EM_ASM is) then it must be a pre-js. This is just JS calling JS, and the pre-js is needed as then it is combined with all the rest of the code and optimized together, so we don't remove code you want.

**Comment by Volkhyn** [comment]
Sorry I badly explained what I tried and pasted the wrong code, test was in EM_ASM. I think that the real issue is coming from the command line because it behaves like if the options like --js-library are not taken in account. I also tried to put directly the js function test in librairy.js and it was working. I also tried to use the embind way to do the same thing, and had an error "missing function : _emval_get_global" as if --bind wasn't there. As I'm using ShiVa3D to compile I will see if the issue is not coming from there.
